pybossa/sched.py

```diff
@@ -590,20 +591,31 @@ def acquire_lock(task_id, user_id, limit, timeout, pipeline=None, execute=True):
     return False
 
 
-def release_reserve_task_lock_by_id(project_id, task_id, user_id, timeout, expiry=EXPIRE_RESERVE_TASK_LOCK_DELAY):
+def release_reserve_task_lock_by_id(project_id, task_id, user_id, timeout, expiry=EXPIRE_RESERVE_TASK_LOCK_DELAY, release_all_task=None):
     reserve_key = get_reserve_task_key(task_id)
     if not reserve_key:
         return
 
     redis_conn = sentinel.master
     lock_manager = LockManager(redis_conn, timeout)
-    resource_id = "reserve_task:project:{}:category:{}:user:{}:task:{}".format(
-        project_id, reserve_key, user_id, task_id)
-    lock_manager.release_reserve_task_lock(resource_id, expiry)
-    current_app.logger.info(
-        "Release reserve task lock. project %s, task %s, user %s, expiry %d",
-        project_id, task_id, user_id, expiry
-    )
+    if release_all_task:
+        pattern = "reserve_task:project:{}:category:{}:user:{}:task:*".format(
+            project_id, reserve_key, user_id)
+        resource_ids = lock_manager.scan_keys(pattern)
+        for k in resource_ids:
+            task_id_in_key = int(k.decode().split(":")[-1])
+            # If a task is locked by the user(in other tab), then the category lock should not be released
+            if task_id_in_key == task_id or not lock_manager.has_lock(get_task_users_key(task_id_in_key), user_id):
```

Comment:
Would it be better to use [`get_user_tasks`](https://github.com/bloomberg/pybossa/blob/1dff9a24d535477100f288c7c2b361b30c75d06b/pybossa/sched.py#L740) func to fetch all tasks that the user has locked, and replace `lock_manager.has_lock(get_task_users_key(task_id_in_key), user_id)`? In this way we only need to query Redis once.

```
tasks_locked_by_user = get_user_tasks(timeout, user_id)

for k in resource_ids:
    task_id_in_key = int(k.decode().split(":")[-1])
    if task_id_in_key == task_id and not task_id_in_key in tasks_locked_by_user:
        # release reserve task lock 
```

Reply:
Yes, this is much better. Will update it. Thanks for the great suggestion.

Reply:
@ywrsusan  Looks like the implementation of ```get_user_tasks``` doesn't release the lock when a task is submitted. I use the code below:
```
        tasks_locked_by_user = get_user_tasks(user_id, timeout)
        print("******* tasks_locked_by_user: ", tasks_locked_by_user)
        for k in resource_ids:
            task_id_in_key = int(k.decode().split(":")[-1])
            # If a task is locked by the user(in other tab), then the category lock should not be released
            if task_id_in_key == task_id or str(task_id_in_key) not in tasks_locked_by_user:
                lock_manager.release_reserve_task_lock(k, expiry)
```
I can see when I submit a task and wait more than 5 seconds, the print statement still print the task id.

Reply:
According to the code, it does release the "get_user_tasks": https://github.com/bloomberg/pybossa/blob/d348769402db653e92ad79d7a95ab8aef6bda522/pybossa/sched.py#L710